syntax = "proto3";
option cc_enable_arenas = true;

import "replication/common/request.proto";

package replication.ir.proto;

message OpID {
    uint64 clientid = 1;
    uint64 clientreqid = 2;
}

// For the view change and recovery protocol, a replica stores two things on
// disk: (1) its current view and (2) the latest view during which it was
// NORMAL. Replicas pack this information into this proto buf and serialize it
// to disk.
message PersistedViewInfo {
  uint64 view = 1;
  uint64 latest_normal_view = 2;
}

enum RecordEntryState {
    RECORD_STATE_TENTATIVE = 0;
    RECORD_STATE_FINALIZED = 1;
}

enum RecordEntryType {
    RECORD_TYPE_INCONSISTENT = 0;
    RECORD_TYPE_CONSENSUS = 1;
}

message RecordEntryProto {
  uint64 view = 1;
  OpID opid = 2;
  RecordEntryState state = 3;
  RecordEntryType type = 4;
  bytes op = 5;
  bytes result = 6;
}

// TODO: Currently, replicas send entire records to one another. Figure out if
// there is a more efficient way to exchange records.
message RecordProto {
  repeated RecordEntryProto entry = 1;
}

message ProposeInconsistentMessage {
    replication.Request req = 1;
}

message ReplyInconsistentMessage {
    uint64 view = 1;
    uint32 replicaIdx = 2;
    OpID opid = 3;
    bool finalized = 4;
}

message FinalizeInconsistentMessage {
    OpID opid = 1;
}

message ConfirmMessage {
    uint64 view = 1;
    uint32 replicaIdx = 2;
    OpID opid = 3;
}

message ProposeConsensusMessage {
    replication.Request req = 1;
}

message ReplyConsensusMessage {
    uint64 view = 1;
    uint32 replicaIdx = 2;
    OpID opid = 3;
    bytes result = 4;
    bool finalized = 5;
}

message FinalizeConsensusMessage {
    OpID opid = 1;
    bytes result = 2;
}

message DoViewChangeMessage {
    uint32 replicaIdx = 1;
    // record is because a replica only sends its record to the
    // leader of the new view.
    RecordProto record = 2;
    uint64 new_view = 3;
    uint64 latest_normal_view = 4;
}
message StartViewMessage {
    RecordProto record = 1;
    uint64 new_view = 2;
}

message UnloggedRequestMessage {
    replication.UnloggedRequest req = 1;
}

message UnloggedReplyMessage {
    bytes reply = 1;
    uint64 clientreqid = 2;
}
